### PR TITLE
RT-DETR

### PR DESCRIPTION
**Pull request by Laughing-q**
@glenn-jocher @AyushExel 
```python
from ultralytics.rtdetr import RTDETRPredictor, RTDETRValidator
from ultralytics.nn.tasks import attempt_load_one_weight
from ultralytics.yolo.utils import DEFAULT_CFG
from ultralytics.yolo.cfg import get_cfg

model = attempt_load_one_weight("rt-detr-l.pt")[0]
# for predict
predictor = RTDETRPredictor(overrides=dict(show=True, mode="predict", half=True))
predictor.setup_model(model=model)
predictor(source="ultralytics/assets/")

# for val
overrides=dict(data="coco.yaml", imgsz=640, mode="val", half=False)
args = get_cfg(cfg=DEFAULT_CFG, overrides=overrides)
validator = RTDETRValidator(args=args)
validator(model=model)
```
- TODO
  - [x] Haven't built top-level api like `YOLO().predict/val` to RT-DETR cause I haven't got a good idea yet. 
  - [x] Also I temporarily put its `predict.py/val.py` in `ultralytics/rtdetr/` which I think probably it's better we create a `ultralytics/transformer/` folder to put transformer-based model like `sam` and `rt-detr`, something like this.
    ```bash
    ├── ultralytics
    │   ├── transformer
    │   │   ├── sam
    │   │   └── rtdetr
    │   ├── yolo
    │   │   ├── v8
    ```
usage like `YOLO`:
```python
from ultralytics.vit import RTDERT
model = RTDERT("rtdetr-l.pt")
# predict
model.predict(source="ultralytics/assets", conf=0.4)
# val
model.val(data="/coco.yaml", half=True)
# export to onnx
model.export(format="onnx", device="0")
```

### Summary
🆕🛠️♻️

This pull request introduces a new model, RT-DETR, based on YOLOv5 and Deformable DETR, that uses multi-scale deformable attention and RepConv blocks for object detection. It adds or modifies several modules in the `ultralytics.nn` package to support the model creation, optimization, and loading processes. It also adds or modifies several files in the `ultralytics.models.rt-detr` and `ultralytics.rtdetr` packages to define the model configuration, prediction, and validation logic.

> _We're building a new model for detection_
> _With `RT-DETR` and `RepConv` sections_
> _We'll import and parse and fuse them all_
> _And heave away, me lads and lasses, heave away_

### Walkthrough
*  Add a new model RT-DETR based on YOLOv5 and Deformable DETR that uses multi-scale deformable attention and RepConv blocks for object detection ([link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-d05b3bd131a1ef119dc8828bf37eb868cdd92ddb2f4d400f0697f75ab644fb33R1-R49), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-6303b21d6bfabc4f1538ddd923a67dc828efd6a7496c04afa5044d6680d477d3R1-R53), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-9cc0562cbdd5ea53281608fcca608bc46bcc52a41f78bbe8ed709f37f1e52590R1-R12), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-231e469c552cef79a698303c0cc0091ee2ca2c494f39b391b63d19f210e9555bR1-R295), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-fcc806c96442cf4cad70250e2d05f2c471251567325095f2c2174fc475b54ea8R1-R271), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-fc9bd8e217a8c71a55895011182403a8a55b9fe8751949a92dc3b9d75d45ce57R1-R389), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-21bf8b929fbb284ba61db2be293d535dd734008479c2a4595ecc6cac659674beR1-R75), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4R340-R355), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4L418-R437), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4L475-R494), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4L481-R516), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-a7ffe2939495acd33e466f6236097180df8034909d20357c5103a34975b9146cR1-R2), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-b9b3a7008e4125fc1ca6181620812b28368e92ffc1218556f68f4fcb8641a423R1-R39), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-aaef13c61a06f92d083a6237aeb7d3475c8493f1b4fbf2ef89e7432218c24da5R1-R108))
*  Add configuration files for the RT-DETR model with different scales ([link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-d05b3bd131a1ef119dc8828bf37eb868cdd92ddb2f4d400f0697f75ab644fb33R1-R49), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-6303b21d6bfabc4f1538ddd923a67dc828efd6a7496c04afa5044d6680d477d3R1-R53))
*  Implement custom postprocessing and pre-transforming logic for the RT-DETR model in the `RTDETRPredictor` and `RTDETRValidator` classes ([link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-b9b3a7008e4125fc1ca6181620812b28368e92ffc1218556f68f4fcb8641a423R1-R39), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-aaef13c61a06f92d083a6237aeb7d3475c8493f1b4fbf2ef89e7432218c24da5R1-R108), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-bd2831d027341da2474db760605211ea00b347396b0d708a4343a2b709eb3675L118-R118), [link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-bd2831d027341da2474db760605211ea00b347396b0d708a4343a2b709eb3675R128-R139))
*  Fuse the convolution and batch normalization layers of the `RepConv` module for faster inference in the `fuse` method of the `BaseModel` class ([link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4L12-R13))
*  Add a default value for the `train_args` key in the `attempt_load_one_weight` function to avoid a KeyError for some older models ([link](https://github.com/ultralytics/ultralytics/pull/2495/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4R108-R110))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Implemented RT-DETR model architecture in the Ultralytics YOLOv8 project.

### 📊 Key Changes
- Deletion of `nn/modules.md` and distribution of its content to dedicated markdown files for blocks, convolutions, heads, transformers, and utils respectively.
- Creation of RT-DETR related files in `vit/rtdetr/` including model loading, prediction, and validation.
- Modifications in existing files like `tasks.py`, `predictor.py`, and `decoders.py` to support RT-DETR model requirements.
- Addition of model names and other changes to support RT-DETR model files in `downloads.py`.

### 🎯 Purpose & Impact
- The modifications support the integration of the RT-DETR model, improving the object detection capabilities of the Ultralytics YOLOv8 project.
- Refactoring of documentation for neural network modules improves maintainability and readability.
- Enhancements in existing codebase facilitate the seamless use of the newly added RT-DETR model by end-users.

### CODE REVIEW

**Comment by AyushExel** [comment]
@Laughing-q @glenn-jocher shold we move rt-detr inside vit folder? That way all transformer/attention based models are will be under one folder

**Comment by glenn-jocher** [comment]
@Laughing-q @AyushExel yeah I agree, vit/ is nice and short too, let's put rt-detr in there next to SAM.

**Comment by Laughing-q** [comment]
@AyushExel @glenn-jocher ok I've added top-level api and move the folder in `vit/` and also fixed a model bug. Now the following  usage is supported:
```python
from ultralytics.vit import RTDERT
model = RTDERT("rtdetr-l.pt")
# predict
model.predict(source="ultralytics/assets", conf=0.4)
# val
model.val(data="/coco.yaml", half=True)
# export to onnx
model.export(format="onnx", device="0")
```

**Comment by AyushExel** [comment]
I think the conflict is because the modules.md is deleted. I tried to merge main but the conflict still didn't resolve

**Comment by glenn-jocher** [comment]
@Laughing-q @AyushExel ok I've merged main into this branch, it's all up to speed. 

Should we wait or try to merge this PR?

**Comment by glenn-jocher** [comment]
@Laughing-q @AyushExel I get these two warnings:

```
/Users/glennjocher/PycharmProjects/ultralytics/venv/lib/python3.10/site-packages/torch/nn/modules/module.py:675: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/build/aten/src/ATen/core/TensorBody.h:485.)
  if param.grad is not None:

/Users/glennjocher/PycharmProjects/ultralytics/venv/lib/python3.10/site-packages/torch/nn/functional.py:1967: UserWarning: nn.functional.sigmoid is deprecated. Use torch.sigmoid instead.
  warnings.warn("nn.functional.sigmoid is deprecated. Use torch.sigmoid instead.")
```

**Comment by Laughing-q** [comment]
@glenn-jocher It's from the fuse operation in repconv, I haven't found a quick way to fix it.

**Comment by glenn-jocher** [comment]
> @glenn-jocher It's from the fuse operation in repconv, I haven't found a quick way to fix it. https://github.com/ultralytics/ultralytics/blob/rt-detr/ultralytics/nn/modules/conv.py#L203-L203
> 
> it looks like it's because we create a new self.conv here.

That should be fine. The real problem is here I think, there's a lot of parameters with gradients, but we should be fusing and running inference without any gradients.

rt-detr-l summary: 494 layers, 32148140 parameters, 25067180 gradients

**Comment by glenn-jocher** [comment]
> @glenn-jocher It's from the fuse operation in repconv, I haven't found a quick way to fix it. https://github.com/ultralytics/ultralytics/blob/rt-detr/ultralytics/nn/modules/conv.py#L203-L203
> 
> it looks like it's because we create a new self.conv here.

That should be fine. The real problem is here I think, there's a lot of parameters with gradients, but we should be fusing and running inference without any gradients.

rt-detr-l summary: 494 layers, 32148140 parameters, 25067180 gradients

**Comment by glenn-jocher** [comment]
> @glenn-jocher It's from the fuse operation in repconv, I haven't found a quick way to fix it. https://github.com/ultralytics/ultralytics/blob/rt-detr/ultralytics/nn/modules/conv.py#L203-L203
> 
> it looks like it's because we create a new self.conv here.

That should be fine. The real problem is here I think, there's a lot of parameters with gradients, but we should be fusing and running inference without any gradients.

rt-detr-l summary: 494 layers, 32148140 parameters, 25067180 gradients

**Comment by Zephyr69** [comment]
Would there be support for training RT-DETR in the future?

**Comment by Laughing-q** [comment]
@Zephyr69 yeah I think we would!

**Comment by xbkaishui** [comment]
hi guys, what's the plan of training with rt-detr ?

**Comment by glenn-jocher** [comment]
@xbkaishui hi there! Thank you for your interest in our work. For training with rt-detr, we are currently exploring different options and plan to provide support for it in the future. Please keep an eye on our GitHub repo for updates. Let us know if you have any other questions or concerns.

**Comment by Adeelbek** [comment]
Hi there (@Laughing-q, @glenn-jocher), 

Thanks for adding new state-of-the-art RT-DETR (Hybrid Convolution and Transformer) model into ```ultralytics``` eco-systems. 
It is always exciting to work with new models. Here, I was wondering if you guys can share your method on how to convert the model from ```.pdparams``` into ```pt``` format. You guys must be doing this somewhere under the hood of ```ultralytics```. 

Thanks again for your work and contribution!

**Comment by glenn-jocher** [comment]
@Adeelbek hello! Thank you for your interest in our work and for your kind words. Regarding converting the RT-DETR model from ```.pdparams``` to ```pt``` format, we recommend using the official conversion tool provided by the PyTorch team. You can find this tool in the PyTorch repository, under the ```tools/convert-tensorflow-to-pytorch``` directory. The usage is quite straightforward, you simply specify the path to your source model file (```.pdparams``` in this case) and the path where you want to save the converted model file (```pt``` format).

As for how we convert the model within the ```ultralytics``` ecosystem, we essentially use this same approach behind the scenes. We do provide some convenience functions to download and convert some models automatically, but in general, we rely on the PyTorch conversion tool to do the actual conversion.

Let us know if you have any other questions or concerns.

**Comment by Adeelbek** [comment]
@glenn-jocher Thanks Glenn for your quick reply. 

Actually, I checked the official repo of PyTorch (from this link https://github.com/pytorch/pytorch/tree/main/tools) but could not find the ```tools/convert-tensorflow-to-pytorch``` directory. Tools does not contain such directory in the official repo of PyTorch. I think ```pdparams``` is yet another deep learning library. So i was wondering if ```convert-tensorflow-to-pytorch``` can also cover the conversion of ```pdparams``` into ```pytorch```. 

It would be very kind of you, if you kindly send me the link of conversion repo that you mentioned above. 

Thanks:-)

**Comment by Adeelbek** [comment]
@Laughing-q Hi, thanks for the idea. I will do the same and hope it works. One last thing to ask. Currently, my ```pdparams``` weights are acquired for the smallest (ResNet18 backbone from here: https://github.com/PaddlePaddle/PaddleDetection/tree/develop/configs/rtdetr) scale of RT-DETR. So if I successfully convert ```pdparams``` weights into ```pt``` format and load it with ```RTDETR``` which was imported from ```from ultralytics import RTDETR```, does it works? I think I have to define the model in the ```./model/rtdetr-resnet18.yaml``` file.  

Thanks again for the method you explained!
Cheers

**Comment by Laughing-q** [comment]
@Adeelbek yeah I think that should work.

**Comment by Adeelbek** [comment]
@Laughing-q ,

I was wondering how you "initialized a pytorch model"? Could u explain bit more? 
thanks again for your reply

**Comment by Laughing-q** [comment]
@Adeelbek Implement the model by torch api and create corresponding yaml file, then create it by:
```python
from ultralytics.nn.tasks import RTDETRDetectionModel
model=RTDETRDetectionModel("model.yaml")
model.load_state_dict(new_dict)
torch.save({"model": model.half()}, f"model.pt")
```
which the `new_dict` here is the params migrated from paddle to torch.

**Comment by WeiChihChern** [comment]
@Laughing-q Is there a plan for building a RT-DERT based instance segmentation model?

**Comment by Laughing-q** [comment]
@WeiChihChern currently I would say "probably", since now the rtdetr training is not as good/easy as yolo models to get good results(the default training strategies might need to be improved). After we make it as easy to get good results as our yolo models, then we might consider that. :)

**Comment by Zephyr69** [comment]
@Laughing-q Should these return values be .long()'d?
https://github.com/ultralytics/ultralytics/blob/d8701b42caeb9f7f1de5fd45e7c3f3cf1724ebb6/ultralytics/vit/utils/loss.py#L165-L170

I'm getting index type error without it.

**Comment by Laughing-q** [comment]
@Zephyr69 the training works well at my side, and I checked the type here it's `long` and `int32`, what' the type in your case? could you provide a reproducible example of your case?

**Comment by Zephyr69** [comment]
> @Zephyr69 the training works well at my side, and I checked the type here it's `long` for `batch_idx, src_idx` and `int32` for `dst_idx`, what' the type in your case? could you provide a reproducible example of your case?

All int32 in my case. I'm on torch1.12.1 on this machine. I'm using a custom dataset in yolo .txt format; I've also tried coco128 and it's the same error.
I may try it on torch2.0 when I have a server instance freed up.

**Comment by Laughing-q** [comment]
@Zephyr69 yeah I'm using torch2.0, but I think I need to review this place of code to make sure it works well with other torch version. :) Thanks for letting me know!
